firedrake/adjoint/variational_solver.py

```diff
@@ -64,3 +77,44 @@ def wrapper(self, **kwargs):
             return out
 
         return wrapper
+
+    @no_annotations
+    def _ad_problem_clone(self, problem, dependencies):
+        """Replaces every coefficient in the residual and jacobian with a deepcopy to return
+        a clone of the original NonlinearVariationalProblem instance. We'll be modifying the
+        numerical values of the coefficients in the residual and jacobian, so in order not to
+        affect the user-defined self._ad_problem.F, self._ad_problem.J and self._ad_problem.u
+        expressions, we'll instead create clones of them.
+        """
+        from firedrake import NonlinearVariationalProblem
+        F_replace_map = {}
+        J_replace_map = {}
+
+        F_coefficients = problem.F.coefficients()
+        J_coefficients = problem.J.coefficients()
+
+        _ad_count_map = {}
+        for block_variable in dependencies:
+            coeff = block_variable.output
+            if coeff in F_coefficients and coeff not in F_replace_map:
+                if isinstance(coeff, Constant):
+                    F_replace_map[coeff] = copy.deepcopy(coeff)
+                else:
+                    F_replace_map[coeff] = coeff.copy(deepcopy=True)
+                _ad_count_map[F_replace_map[coeff]] = coeff.count()
+
+            if coeff in J_coefficients and coeff not in J_replace_map:
+                if coeff in F_replace_map:
+                    J_replace_map[coeff] = F_replace_map[coeff]
+                elif isinstance(coeff, Constant):
+                    J_replace_map[coeff] = copy.deepcopy(coeff)
+                else:
+                    J_replace_map[coeff] = coeff.copy()
+                _ad_count_map[J_replace_map[coeff]] = coeff.count()
+
+        nlvp = NonlinearVariationalProblem(replace(problem.F, F_replace_map),
+                                           F_replace_map[problem.u],
+                                           bcs=problem.bcs,
+                                           J=replace(problem.J, J_replace_map))
+        nlvp._ad_count_map = _ad_count_map
```

Comment:
Similarly could we also define/set up the `_ad_count_map` attribute inside the class (i.e. inside the mixin)? So that we don't end up defining a new attribute out of the class.